### PR TITLE
Honor podLabelSelector for KV-event subscriptions

### PR DESCRIPTION
**Pull request by vMaroon**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Apply `kvEventsConfig.podDiscoveryConfig.podLabelSelector` to endpoints delivered by `endpoint-notification-source`. Pool members that publish no KV events, such as decode-only pods in a P/D pool, otherwise get dialed and retried indefinitely. On a 32-prefill/12-decode GLM-5.2 pool, each EPP replica logged about 2.5 subscriber connection-refused errors per second (48k in six hours), burying real errors.

Parse the selector at startup when pod discovery is enabled and subscribe only to matching endpoints. The default selector is empty, so omitted or empty selectors match every endpoint; an invalid explicit selector fails startup. When an endpoint stops matching, remove its subscriber and cached index entries. Excluded add/update events clear cached entries only when a subscriber was removed. Deletion performs cleanup even without a subscriber, since speculative entries may exist.

The original in-cluster validation of the selector filter reduced subscriber connection errors to zero on that pool, with index admissions and lookups unchanged. The lifecycle cleanup changes were verified locally.

**Which issue(s) this PR fixes**:

None; localized bug fix.

**Test plan**:

- [x] Configuration controls subscriptions for equality, set-based, and empty selectors; invalid selectors fail startup only when discovery is enabled.
- [x] Label changes and deletions remove subscribers and cached entries; endpoints can subscribe again when their labels match.
- [x] Omitted parameters and nested discovery configs preserve subscriptions for unlabeled endpoints.
- [x] Excluded add/update events skip index cleanup when no subscriber exists; deletion still clears cached entries.

**Release note**:

```release-note
With `kvEventsConfig.discoverPods: true`, `precise-prefix-cache-producer` applies `kvEventsConfig.podDiscoveryConfig.podLabelSelector` to endpoint notifications. Only matching endpoints receive KV-event subscriptions. The selector defaults to empty: omitted or empty selectors include all endpoints, preserving subscriptions for existing configs that leave it unset. Invalid selectors fail EPP startup when pod discovery is enabled. Endpoints that stop matching have their subscriptions and cached index entries removed.
```

### CODE REVIEW

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go [hunk 1]
```diff
@@ -158,6 +160,16 @@ func PluginFactory(name string, rawParameters *json.Decoder, handle plugin.Handl
 // The kvcache indexer, KV-events pool, and any local ZMQ subscriber start
 // in background goroutines bound to ctx.
 func New(ctx context.Context, name string, config PluginConfig) (*Producer, error) {
+	var podSelector labels.Selector
+	if kc := config.KVEventsConfig; kc != nil && kc.DiscoverPods && kc.PodDiscoveryConfig != nil && kc.PodDiscoveryConfig.PodLabelSelector != "" {
```
Comment (bongwoobak) [inline pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go:164]: Nothing on main actually reads `podLabelSelector` today, this PR is the first consumer, and the default in `kvevents.DefaultConfig()` is `llm-d.ai/inference-serving=true`, not empty. `PluginFactory` starts from that default and merges the user's JSON on top, so anyone who never set the field starts filtering on upgrade and pods without that label quietly lose their subscriptions. The test sets the field explicitly in every case so the default path isn't covered.

Can the default become `""` here? Or at least a release note line saying what happens to configs that leave it unset.

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go [hunk 2]
```diff
@@ -44,13 +44,14 @@ func (p *Producer) Extract(ctx context.Context, event fwkdl.EndpointEvent) error
...
+	case event.Type == fwkdl.EventAddOrUpdate && matchesSelector:
 		if err := p.ensureSubscriber(ctx, meta); err != nil {
 			return err
 		}
 		logger.V(logging.DEBUG).Info("Adding subscriber", "endpoint", endpointKey)
-	case fwkdl.EventDelete:
+	case event.Type == fwkdl.EventAddOrUpdate || event.Type == fwkdl.EventDelete:
 		p.subscribersManager.RemoveSubscriber(ctx, endpointKey)
 		if meta.Address != "" {
 			if err := p.kvCacheIndexer.KVBlockIndex().Clear(ctx, fmt.Sprintf("%s:%s", meta.Address, meta.Port)); err != nil {
```
Comment (bongwoobak) [inline pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go:57]: This also runs on every update of an endpoint the selector excludes, and on Redis `Clear` is a full keyspace SCAN. Those endpoints never had a subscriber, so there is nothing to clear there. If `RemoveSubscriber` returned whether it actually removed something, the clear could be gated on that and the excluded pods stay free.

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go (hunk 1 above)
Comment (elevran) [reply pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go:164]: +1
Should enable backward compatibility or at least have a release note to indicate the new behavior.

**Review by elevran — COMMENTED** [review]
The test additions cover both the startup validation and the lifecycle-cleanup edge cases well.
Backward compat should be considered per @bongwoobak feedback.

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go (hunk 1 above)
Comment (vMaroon) [reply pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/producer.go:164]: Changed the default selector to `""` so configs that omit the field keep subscribing to all discovered endpoints. Added factory-path regression coverage for omitted parameters and nested configs, while explicit selectors still filter.

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go (hunk 2 above)
Comment (vMaroon) [reply pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go:58]: `RemoveSubscriber` now reports whether it removed a subscriber. Excluded add/update events skip `Clear` when it returns false. Deletions still clear entries because speculative indexing can populate them without a subscriber. The regression test verifies repeated excluded updates cause no extra clears.

File: pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go [hunk 3]
```diff
@@ -44,13 +44,14 @@ func (p *Producer) Extract(ctx context.Context, event fwkdl.EndpointEvent) error
...
 
-	switch event.Type {
-	case fwkdl.EventAddOrUpdate:
+	matchesSelector := p.podSelector == nil || p.podSelector.Matches(labels.Set(meta.Labels))
+	switch {
+	case event.Type == fwkdl.EventAddOrUpdate && matchesSelector:
 		if err := p.ensureSubscriber(ctx, meta); err != nil {
 			return err
 		}
 		logger.V(logging.DEBUG).Info("Adding subscriber", "endpoint", endpointKey)
```
[the remark this answers is not in the document]
Comment (vMaroon) [reply pkg/epp/framework/plugins/requestcontrol/dataproducer/preciseprefixcache/extractor.go]: Removed the unconditional log. `ensureSubscriber` already logs after ensuring a subscriber and returns before that log when the address is missing.

**Review by elevran — APPROVED** [review]
Verified the backward-compat fix - thanks.
All three review threads look addressed - approving.
